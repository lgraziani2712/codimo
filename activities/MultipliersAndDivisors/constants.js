/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * @flow
 */

export const LEAVE_MAZE = 'leave_maze';
export const LEAVE_MAZE_TITLE = 'üï≥üèÉüí® Saltar';
export const LEAVE_MAZE_TOOLTIP =
  'SI EL N√öMERO EST√Å EN EL FINAL DEL LABERINTO, SALTA A LA RECTA NUM√âRICA';

export const ACTOR_PROCESS_START = 'actor_process_start';
export const ACTOR_PROCESS_STOP = 'actor_process_stop';

export const ACTOR_PROCESSOR_TITLE_1 = 'EL N√öMERO';
export const ACTOR_PROCESSOR_TITLE_2 = 'SE MUEVE AS√ç:';
export const ACTOR_PROCESSOR_TOOLTIP = 'ELEG√ç EL N√öMERO QUE QUER√âS MOVER';
