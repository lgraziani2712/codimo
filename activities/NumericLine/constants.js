/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * @flow
 */

export const LEAVE_MAZE = 'leave_maze';
export const LEAVE_MAZE_TITLE = '🕳🏃💨 Saltar';
export const LEAVE_MAZE_TOOLTIP =
  'Si el número está sobre una salida, salta hacia la línea numérica';
