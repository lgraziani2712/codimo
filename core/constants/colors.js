/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * @flow
 */

const WHITE = '#fff';

export const ACTION = '#a55b6d';
export const END_ACTION = '#a5a55b';
export const CONTROL = '#5b67a5';
export const CONTAINER = '#369836';

export const COLOR_PALETTE = {
  blue: {
    normal: '#428bca',
    dark: '#0a5a9c',
    color: WHITE,
  },
  green: {
    normal: '#5cb85c',
    dark: '#1b7c1b',
    color: WHITE,
  },
  red: {
    normal: '#d9534f',
    dark: '#be2f2b',
    color: WHITE,
  },
  orange: {
    clear: '#ffc697',
    light: '#f1a667',
    normal: '#d3823f',
    dark: '#b36320',
    color: WHITE,
    transparent: {
      clear: 'rgba(255, 198, 151, 0.4)',
    },
  },
};
