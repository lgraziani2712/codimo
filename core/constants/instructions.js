/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * @flow
 */

export const SIMPLE_LOOP = 'simple_loop';
export const REPEAT_PARAM = 'repeat_param';

///////////////////////////////////////////////////
// Component instructions
///////////////////////////////////////////////////
export const MOVE_FORWARD = 'move_forward';
export const MOVE_RIGHT = 'move_right';
export const MOVE_BACKWARD = 'move_backward';
export const MOVE_LEFT = 'move_left';
