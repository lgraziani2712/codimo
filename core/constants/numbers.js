/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * @flow
 */

export const ZERO = 0;
export const ONE = 1;
export const TWO = 2;
export const FOUR = 4;
export const TEN = 10;

export const HALF = TWO;

export const DEFAULT_MOVEMENT_DURATION = 0.5;
export const ANCHOR_CENTER = 0.5;
