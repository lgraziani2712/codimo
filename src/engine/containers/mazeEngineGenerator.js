/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * @flow
 */
import { Container } from 'pixi.js';

import { ZERO, ONE, HALF } from 'constants/numbers';
import { type Instructions } from 'blockly/executorGenerator';
import {
  MOVE_FORWARD,
  MOVE_RIGHT,
  MOVE_BACKWARD,
  MOVE_LEFT,
  LEAVE_MAZE,
} from 'constants/instructions';
import mazeGenerator, { type MazeData, type Maze } from 'engine/components/mazeGenerator';
import numberGenerator, {
  START_STATE,
  STOP_STATE,
  type NumberActor,
} from 'engine/components/numberGenerator';
import numericLineGenerator, {
  type NumericLineData,
  type NumericLine,
} from 'engine/components/numericLineGenerator';
import {
  MazeExitError,
  MazePathError,
  MazePathOverflowError,
  MazeStarvationError,
  MazeWrongExitError,
} from 'engine/helpers/errors';
import randomizeActorsConfig from 'engine/helpers/randomizeActorsConfig';
import { getRandomInt } from 'helpers/randomizers';

const numberHasLeftMazeConfig = (
  mazeData: MazeData,
  numericLineData: NumericLineData,
  numericLine: NumericLine,
  number: NumberActor,
) => async (): Promise<void> => {
  const exitIdx = mazeData.exits.indexOf(number.position);

  if (exitIdx === -ONE) {
    throw MazeExitError;
  }

  const exit = mazeData.exits[exitIdx];

  await numericLine.receiveNumberAtPosition(number, numericLineData.accesses[exitIdx]);

  if (exit !== number.finalPosition) {
    number.beSad(START_STATE);
    numericLine.beSad(START_STATE);

    throw MazeWrongExitError;
  }

  number.beHappy(START_STATE);
  numericLine.beHappy(START_STATE);
};
const directions = {
  [MOVE_FORWARD]: [ZERO, -ONE],
  [MOVE_RIGHT]: [ONE, ZERO],
  [MOVE_BACKWARD]: [ZERO, ONE],
  [MOVE_LEFT]: [-ONE, ZERO],
};
const directionsToWalls = {
  [MOVE_FORWARD]: 'top',
  [MOVE_RIGHT]: 'right',
  [MOVE_BACKWARD]: 'bottom',
  [MOVE_LEFT]: 'left',
};

const excecuteSetOfInstructionsConfig = (
  mazeData: MazeData,
  number: NumberActor,
  numberHasLeft: () => Promise<void>,
  /**
   * This is the engine's main function. It will execute every instruction and
   * animate everything according to the result of each of one of them.
   *
   * @param {Instructions} instructions  map of instructions
   * @return {Promise<void>}             it will finish or throw an exeption
   */
) => async (instructions: Instructions): Promise<void> => {
  for (let i = 0; i < instructions.length; i++) {
    if (instructions[i] === LEAVE_MAZE) {
      await numberHasLeft();

      break;
    }
    const direction = instructions[i];
    const oldPosition = number.position;
    const path = mazeData.path.get(oldPosition);
    const newPosition =
      oldPosition
          .split(',')
          .map((pos, idx) => (parseInt(pos) + directions[direction][idx]))
          .join(',');

    if (!path || !path[directionsToWalls[direction]]) {
      await number.hitTheWall(directionsToWalls[direction]);

      throw MazePathError;
    }
    /**
     * What happend if the number is over an exit and it tries to go forward?
     * Since there will be an open wall, the engine will try to move it.
     * But it won't enter the numeric line, it will fall from the maze.
     * And we don't want that.
     */
    if (mazeData.exits.indexOf(oldPosition) !== -ONE && direction === MOVE_FORWARD) {
      await number.beTheFallenOne();

      throw MazePathOverflowError;
    }
    await number.updatePosition(newPosition);
  }
  /**
   * If the `position` is not undefined it means
   * it never left the maze.
   */
  if (number.position) {
    throw MazeStarvationError;
  }
};
const handleResetGameConfig = (
  randomizeActor: () => Array<number>,
  number: NumberActor,
  maze: Maze,
  actorExitIdx: number,
  numericLine: NumericLine,
) => () => {
  const newActor = randomizeActor();

  number.changeActor(newActor[actorExitIdx]);
  number.view.setParent(maze.view);
  number.beHappy(STOP_STATE);
  number.beSad(STOP_STATE);
  number.resetPosition();

  numericLine.beHappy(STOP_STATE);
  numericLine.beSad(STOP_STATE);
};

export type GameDifficulty = 'easy' | 'normal' | 'hard';
export type Engine = {|
  view: Container,
  excecuteSetOfInstructions(instructions: Instructions): Promise<void>,
  handleResetGame(): void,
|};
export default function mazeEngineGenerator(
  mazeData: MazeData,
  numericLineData: NumericLineData,
  difficulty: GameDifficulty,
): Engine {
  const view = new Container();
  const numericLine = numericLineGenerator(
    numericLineData.statics,
    mazeData.size,
    mazeData.margin,
  );

  const maze = mazeGenerator(mazeData);
  const randomizeActor = randomizeActorsConfig(
    numericLineData.statics,
    numericLineData.accesses,
    difficulty,
  );
  const randomActor = randomizeActor();
  const actorExitIdx =
    mazeData.actorExitIdx !== undefined
      ? mazeData.actorExitIdx
      : getRandomInt(ZERO, randomActor.length);
  const number = numberGenerator(
    randomActor[actorExitIdx],
    mazeData.access,
    mazeData.exits[actorExitIdx],
    mazeData.size,
    mazeData.margin,
  );
  const MARGIN_NUMERIC_LINE_MAZE = mazeData.margin + mazeData.size / HALF;

  maze.view.addChild(number.view);

  maze.view.y = numericLine.view.height + MARGIN_NUMERIC_LINE_MAZE;

  view.addChild(maze.view, numericLine.view);

  return {
    view,
    excecuteSetOfInstructions: excecuteSetOfInstructionsConfig(
      mazeData, number, numberHasLeftMazeConfig(mazeData, numericLineData, numericLine, number),
    ),
    handleResetGame: handleResetGameConfig(
      randomizeActor, number, maze, actorExitIdx, numericLine,
    ),
  };
}
