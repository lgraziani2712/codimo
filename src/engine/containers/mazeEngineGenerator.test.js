/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * @flow
 */
/* eslint-disable no-magic-numbers */
import gameMetadataDataWithoutBlocks from 'test/gameMetadataDataWithoutBlocks.json';
import * as actionNames from 'constants/instructions';

import mazeEngineGenerator from './mazeEngineGenerator';

const { mazeData, numericLineData, difficulty } = gameMetadataDataWithoutBlocks;

mazeData.path = new Map(mazeData.path);

describe('engine > containers > mazeEngineGenerator', () => {
  it('should parse an array of correct `instructions` and return a response', async () => {
    const mazeEngine = mazeEngineGenerator(mazeData, numericLineData, difficulty);
    const instructions = [
      actionNames.MOVE_FORWARD,
      actionNames.MOVE_FORWARD,
      actionNames.LEAVE_MAZE,
    ];

    try {
      await mazeEngine.excecuteSetOfInstructions(instructions);
    } catch (error) {
      expect(error.name).toBe('MazeExitError');
      expect(error.text).toMatchSnapshot();

      return;
    }
    expect('true').toBe('not evaluated');
  });
  it('should throw a MazePathError if a wall is between valid paths', async () => {
    const newMazeData = {
      ...mazeData,
      path: new Map([['0,0', {}], ['1,0', {}]]),
      access: '0,0',
      exits: ['1,0'],
      actorExitIdx: 0,
    };
    const newNumericLineData = {
      statics: [null, 6, 9],
      accesses: [0],
    };
    const mazeEngine = mazeEngineGenerator(newMazeData, newNumericLineData, difficulty);
    const instructions = [
      actionNames.MOVE_RIGHT,
    ];

    try {
      await mazeEngine.excecuteSetOfInstructions(instructions);
    } catch (error) {
      expect(error.name).toBe('MazePathError');
      expect(error.text).toMatchSnapshot();

      return;
    }
    expect('true').toBe('not evaluated');
  });
  it('should be possible to create a maze with 1-actor:n-exits', async () => {
    const newMazeData = {
      ...mazeData,
      path: new Map([['0,0', {}], ['1,0', {}]]),
      access: '0,0',
      exits: ['1,0'],
      actorExitIdx: 0,
    };
    const newNumericLineData = {
      statics: [null, 6, null],
      accesses: [0, 1],
    };
    const mazeEngine = mazeEngineGenerator(newMazeData, newNumericLineData, difficulty);

    expect(mazeEngine).not.toBeUndefined();
  });
  it('should throw a MazeWrongExitError if the actor leaves at the wrong exit', async () => {
    const newMazeData = {
      ...mazeData,
      path: new Map([
        ['0,0', { right: true }],
        ['1,0', { left: true }],
      ]),
      access: '0,0',
      exits: ['1,0', '1,1'],
      actorExitIdx: 1,
    };
    const newNumericLineData = {
      statics: [null, 6, null],
      accesses: [0, 1],
    };
    const mazeEngine = mazeEngineGenerator(newMazeData, newNumericLineData, difficulty);
    const instructions = [
      actionNames.MOVE_RIGHT,
      actionNames.LEAVE_MAZE,
    ];

    try {
      await mazeEngine.excecuteSetOfInstructions(instructions);
    } catch (error) {
      expect(error.name).toBe('MazeWrongExitError');
      expect(error.text).toMatchSnapshot();

      return;
    }
    expect('true').toBe('not evaluated');
  });
  it('should throw a MazeStarvationError if the actor doesn\'t leaves', async () => {
    const newMazeData = {
      ...mazeData,
      path: new Map([
        ['0,0', { right: true }],
        ['1,0', { left: true }],
      ]),
      access: '0,0',
      exits: ['1,0', '1,1'],
      actorExitIdx: 1,
    };
    const newNumericLineData = {
      statics: [null, 6, null],
      accesses: [0, 1],
    };
    const mazeEngine = mazeEngineGenerator(newMazeData, newNumericLineData, difficulty);
    const instructions = [
      actionNames.MOVE_RIGHT,
    ];

    try {
      await mazeEngine.excecuteSetOfInstructions(instructions);
    } catch (error) {
      expect(error.name).toBe('MazeStarvationError');
      expect(error.text).toMatchSnapshot();

      return;
    }
    expect('true').toBe('not evaluated');
  });
});
