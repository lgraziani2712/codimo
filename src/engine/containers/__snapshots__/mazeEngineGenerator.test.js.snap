// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`engine > containers > mazeEngineGenerator should parse an array of correct \`instructions\` and return a response 1`] = `"El número solo tiene fuerzas para saltar cuando llega a una salida"`;

exports[`engine > containers > mazeEngineGenerator should throw a MazePathError if a wall is between valid paths 1`] = `"La dirección que intenta ir el número es incorrecta"`;

exports[`engine > containers > mazeEngineGenerator should throw a MazeStarvationError if the actor doesn't leaves 1`] = `undefined`;

exports[`engine > containers > mazeEngineGenerator should throw a MazeWrongExitError if the actor leaves at the wrong exit 1`] = `"¿El número es más grande que el de su izquierda y más chico que el de su derecha?"`;
